src/parser.cpp

```diff
@@ -0,0 +1,637 @@
+                    try
+                    {
+                        offset = std::stod(attribute->getValue());
+                    } catch (std::exception) {
+                        UnitsErrorPtr err = std::make_shared<UnitsError>();
+                        err->setDescription("Offset attribute value '" + attribute->getValue() +
+                                            "' in unit '" + childNode->getAttribute("units") +
+                                            "' from units '" + units->getName() + "' is not convertible to decimal number.");
+                        err->setUnits(units);
+                        addError(err);
+                    }
+                } else {
+                    UnitsErrorPtr err = std::make_shared<UnitsError>();
+                    err->setDescription("Invalid attribute '" + attribute->getType() +
+                                        "' found in unit '" + childNode->getAttribute("units") +
+                                        "' from units '" + units->getName() + "'.");
+                    err->setUnits(units);
+                    addError(err);
+                }
+                attribute = attribute->getNext();
+            }
+            units->addUnit(name, prefix, exponent, multiplier, offset);
+        } else {
+            UnitsErrorPtr err = std::make_shared<UnitsError>();
+            err->setDescription("Invalid child element '" + childNode->getType() +
+                                "' found in units '" + units->getName() + "'.");
+            err->setUnits(units);
+            addError(err);
+        }
+        childNode = childNode->getNext();
+    }
+}
+
+void Parser::loadVariable(const VariablePtr &variable, const XmlNodePtr &node)
+{
+    XmlAttributePtr attribute = node->getRootAttribute();
+    while (attribute) {
+        if (attribute->isType("name")) {
+            variable->setName(attribute->getValue());
+        } else if (attribute->isType("units")) {
+            variable->setUnits(attribute->getValue());
+        } else if (attribute->isType("interface")) {
+            variable->setInterfaceType(attribute->getValue());
+        } else if (attribute->isType("initial_value")) {
+            variable->setInitialValue(attribute->getValue());
+        } else {
+            VariableErrorPtr err = std::make_shared<VariableError>();
+            err->setDescription("Invalid attribute '" + attribute->getType() +
+                                "' found in variable '" + node->getAttribute("name") + "'.");
+            err->setVariable(variable);
+            addError(err);
+        }
+        attribute = attribute->getNext();
+    }
+}
+
+void Parser::loadConnection(const ModelPtr &model, const XmlNodePtr &node)
+{
+    // Initialise connection components and variables.
+    ComponentPtr component1 = nullptr;
+    ComponentPtr component2 = nullptr;
+    VariablePtr variable1 = nullptr;
+    VariablePtr variable2 = nullptr;
```

Comment:
Nitpicking (?): why define those variables here when they may not be used at all? I would define them only when we know they are going to be used.